service/controller/clusterapi/v29/resource/clusterazs/create.go

```diff
@@ -63,33 +63,48 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d MachineDeployments for tenant cluster", len(machineDeployments)))
 	}
 
-	var azs map[string]subnetPair
+	// As a first step we initialize the mappings for all the relevant
+	// availability zone information coming from the master node, the worker nodes
+	// and the EC2 subnets. Having all the availability zones inside the map makes
+	// it easier to fill their corresponding subnet and route table information
+	// below.
+	azMapping := map[string]mapping{}
 	{
-		// Acquire AZs together with corresponding subnets from AWS EC2 API
-		// results.
-		azs, err = fromEC2SubnetsToMap(cc.Status.TenantCluster.TCCP.Subnets)
+		azMapping[key.MasterAvailabilityZone(cr)] = mapping{}
+
+		for _, md := range machineDeployments {
+			for _, az := range key.WorkerAvailabilityZones(md) {
+				azMapping[az] = mapping{}
+			}
+		}
+
+		for _, az := range azsFromSubnets(cc.Status.TenantCluster.TCCP.Subnets) {
+			azMapping[az] = mapping{}
+		}
+	}
+
+	// Map subnet, route table and natgateway information to their corresponding availability
+	// zones based on the AWS API results.
+	{
+		azMapping, err = mapNATGateways(azMapping, cc.Status.TenantCluster.TCCP.NATGateways)
 		if err != nil {
 			return microerror.Mask(err)
 		}
 
-		// Include master's AZ if missing.
-		if _, exists := azs[key.MasterAvailabilityZone(cr)]; !exists {
-			azs[key.MasterAvailabilityZone(cr)] = subnetPair{}
+		azMapping, err = mapRouteTables(azMapping, cc.Status.TenantCluster.TCCP.RouteTables)
```

Comment:
I think this is actually no longer needed. We can take a closer look later, so no blocker.